### ISSUE TITLE
ansible_become cannot be overriden by host or task. local_action and delegate_to both affected

### ISSUE THREAD

**Issue by caylorme** [ansible/ansible#12577]
_Playbook.yml_

``` YAML

---
- name: Bug Report Example
  hosts: testnode.testdomain
  tasks:
    - name: Create lib directory
      file: path="lib/" state=directory
      #local_action: file path="lib/" state=directory
      become: no
      delegate_to: localhost
    - name: Can't do this...
      #local_action: template src=text.j2 dest=lib/file.txt
      template: src=text.j2 dest=lib/file.txt
      delegate_to: localhost
      become: no
```

_inventory.ini_

``` ini
localhost   ansible_connection=local ansible_become=no

testnode.testdomain ansible_become=yes

```

_text.j2_

```
# {{ ansible_managed }}
Hello World!
```

_Results_

```
PLAY [Bug Report Example] *****************************************************

GATHERING FACTS ***************************************************************
ok: [testnode.testdomain]

TASK: [Create lib directory] **************************************************
failed: [testnode.testdomain -> localhost] => {"failed": true, "parsed": false}
[sudo via ansible, key=frbjmfdvrioeeqvxbfxiqszlclvpqwjt] password:


FATAL: all hosts have already failed -- aborting
```

I can swap out the delegate_to with the local_action and get the same results.
If I remove the ansible_become definition from the originating node it will succeed.

Related Issue:
https://github.com/ansible/ansible/issues/10906

**Comment by caylorme** [comment]
@jimi-c and @bcoca, thanks for taking this, is there any update?
Is there an obvious cause for this?

**Comment by jimi-c** [comment]
Hi @caylorme, I just tested this using your example and I'm not seeing a failure any more:

<pre>
PLAY [Bug Report Example] ******************************************************
TASK [setup] *******************************************************************
ok: [testnode.testdomain]
TASK [Create lib directory] ****************************************************
ok: [testnode.testdomain -> localhost] => {"changed": false, "gid": 0, "group": "root", "mode": "0755", "owner": "root", "path": "lib/", "secontext": "unconfined_u:object_r:admin_home_t:s0", "size": 4096, "state": "directory", "uid": 0}
TASK [Can't do this...] ********************************************************
ok: [testnode.testdomain -> localhost] => {"changed": false, "gid": 0, "group": "root", "mode": "0644", "owner": "root", "path": "lib/file.txt", "secontext": "unconfined_u:object_r:admin_home_t:s0", "size": 108, "state": "file", "uid": 0}
PLAY RECAP *********************************************************************
testnode.testdomain        : ok=3    changed=0    unreachable=0    failed=0   
</pre>

Can you confirm if you're still seeing this on your end?

**Comment by caylorme** [comment]
@jimi-c, you must be running as root?
I'm experiencing this while running ansible as a local user (not root), who requires password to sudo. The remote user on its remote host does not require sudo password. I want all actions on the remote host to use SUDO, but the delegate_to: localhost AND local_action tasks need to run whatever I have set for localhost or what I have explicitly defined for that task (become: no)

**Comment by caylorme** [comment]
I'm thinking you might be running ansible as a user who does not require a sudo password. Or are already authorized to sudo.

**Comment by jimi-c** [comment]
@caylorme nope, I've also tested with a non-root user and it works for me there too. With which version of Ansible are you seeing this problem? I'm running on the latest version of devel.

**Comment by caylorme** [comment]
Just tried it with 1.9.4, still fails.

$ ansible --version
ansible 1.9.4
  configured module search path = None

**Comment by jimi-c** [comment]
Can you test with devel in anwyay to see if the problem has been resolved there, as it appears?

**Comment by caylorme** [comment]
Sure, though I do remember doing this at some point.

**Comment by caylorme** [comment]
Seems to work on latest dev:

```
$ ansible --version
ansible 2.0.0 (devel 4486e136ea) last updated 2015/10/26 14:40:33 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 06f301b05b) last updated 2015/10/26 14:40:55 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 405c3cb2c7) last updated 2015/10/26 14:41:00 (GMT -400)
  config file =
  configured module search path = None

$ ansible-playbook -i inventory.ini playbook.yml

PLAY [Bug Report Example] ******************************************************

TASK [setup] *******************************************************************
ok: [testnode.testdomain]

TASK [Create lib directory] ****************************************************
changed: [testnode.testdomain -> localhost]

TASK [Can't do this...] ********************************************************
changed: [testnode.testdomain] -> localhost]

PLAY RECAP *********************************************************************
testnode.testdomain              : ok=3    changed=2    unreachable=0    failed=0
```

**Comment by jimi-c** [comment]
Ok, I've moved this off the v2 milestone and to the stable-1.9 milestone, in case this is something we want to look at fixing/backporting. Most likely we won't however, due to the fact that there is a work-around here as you noted above.

**Comment by diefans** [comment]
I have a similar problem with ansible==2.0.0.2.

I want to make some local actions without becoming root within a block:

```
- become: no
  delegate_to: localhost
  block:
    - command: hostname
```

...which ends up in:
`fatal: [test -> localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "command"}, "module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}`

One way to fix this is by setting the fact of `ansible_become=no`and after the block reset this value, but the whole thing is quite expressive:

```
- delegate_to: localhost
  block:
    - set_fact:
        old_ansible_become: "{{ ansible_become }}"

    - set_fact:
        ansible_become: no

    - command: hostname

  always:
    - set_fact:
        ansible_become: "{{ old_ansible_become }}"
```

the other way would be to set the host var for localhost `ansible_become=no`, but this seems wrong.

**Comment by bcoca** [comment]
So i believe #15024 should fix this and apply the correct become settings (either task ones or delegated host ones) and ignore the 'inventory_hostname' ones

**Comment by bcoca** [comment]
closing via merged #15024
